File: serializer/src/main/scala/com/ning/serializer/usescala/Serializer.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ning.serializer.usescala

import java.io._
import java.nio.ByteBuffer

import scala.reflect.ClassTag

/**
  * 序列化工具类 主要负责返回序列化实例 类似Java工厂类
  */
abstract class Serializer {
  //默认类加载器
  @volatile protected var defaultClassLoader: Option[ClassLoader] = None

  /**
   * Sets a class loader for the serializer to use in deserialization.
   *
   * @return this Serializer object
   */
  def setDefaultClassLoader(classLoader: ClassLoader): Serializer = {
    defaultClassLoader = Some(classLoader)
    this
  }

  /** Creates a new [[SerializerInstance]]. */
  def newInstance(): SerializerInstance

}


object Serializer {
  def getSerializer(serializer: Serializer): Serializer = {
      serializer
  }

  def getSerializer(serializer: Option[Serializer]): Serializer = {
    serializer.get
  }
}


/**
 * :: DeveloperApi ::
 * An instance of a serializer, for use by one thread at a time.
 *
 * It is legal to create multiple serialization / deserialization streams from the same
 * SerializerInstance as long as those streams are all used within the same thread.
  *
  * 序列化实例
  * 定义实例的抽象方法 包括序列化对象，反序列化对象 将输出流包装为序列化流 将输入流包装为序列化流
 */
abstract class SerializerInstance {
  def serialize[T: ClassTag](t: T): ByteBuffer

  def deserialize[T: ClassTag](bytes: ByteBuffer): T

  def deserialize[T: ClassTag](bytes: ByteBuffer, loader: ClassLoader): T

  def serializeStream(s: OutputStream): SerializationStream

  def deserializeStream(s: InputStream): DeserializationStream
}

/**
 * :: DeveloperApi ::
 * A stream for writing serialized objects.
  *
  * 序列化流  负责将对象写出到序列化流
  * */
abstract class SerializationStream {
  /** The most general-purpose method to write an object. */
  def writeObject[T: ClassTag](t: T): SerializationStream
  /** Writes the object representing the key of a key-value pair. */
  def writeKey[T: ClassTag](key: T): SerializationStream = writeObject(key)
  /** Writes the object representing the value of a key-value pair. */
  def writeValue[T: ClassTag](value: T): SerializationStream = writeObject(value)
  def flush(): Unit
  def close(): Unit

  def writeAll[T: ClassTag](iter: Iterator[T]): SerializationStream = {
    while (iter.hasNext) {
      writeObject(iter.next())
    }
    this
  }
}


/**
 * :: DeveloperApi ::
 * A stream for reading serialized objects.
  * 反序列化抽象 负责将对象反序列化为JAVA对象
 */
abstract class DeserializationStream {
  /** The most general-purpose method to read an object. */
  def readObject[T: ClassTag](): T
  /** Reads the object representing the key of a key-value pair. */
  def readKey[T: ClassTag](): T = readObject[T]()
  /** Reads the object representing the value of a key-value pair. */
  def readValue[T: ClassTag](): T = readObject[T]()
  def close(): Unit

  /**
   * Read the elements of this stream through an iterator. This can only be called once, as
   * reading each element will consume data from the input source.
    *
   */
  def asIterator: Iterator[Any] = new NextIterator[Any] {
    override protected def getNext() = {
      try {
        readObject[Any]()
      } catch {
        case eof: EOFException =>
          finished = true
          null
      }
    }

    override protected def close() {
      DeserializationStream.this.close()
    }
  }

  /**
   * Read the elements of this stream through an iterator over key-value pairs. This can only be
   * called once, as reading each element will consume data from the input source.
   */
  def asKeyValueIterator: Iterator[(Any, Any)] = new NextIterator[(Any, Any)] {
    override protected def getNext() = {
      try {
        (readKey[Any](), readValue[Any]())
      } catch {
        case eof: EOFException => {
          finished = true
          null
        }
      }
    }

    override protected def close() {
      DeserializationStream.this.close()
    }
  }
}

File: serializer/src/main/scala/com/ning/serializer/usejava/Serializer.java
package com.ning.serializer.usejava;


import java.io.EOFException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.util.Iterator;
import java.util.Optional;

/**
 * Created by zhaoshufen
 * User:  zhaoshufen
 * Date: 2017/9/29
 * Time: 21:18
 * To change this setting on:Preferences->editor->File and Code Templates->Include->File Header
 */
public abstract class Serializer {
    //默认类加载器
    protected Optional<ClassLoader> defaultClassLoader = Optional.empty();
    public void setDefaultClassLoader( Optional<ClassLoader> defaultClassLoader ){
        this.defaultClassLoader = defaultClassLoader;
    }
    public abstract SerializerInstance newInstance();

}
abstract class SerializerInstance {
    public  abstract <T> ByteBuffer serialize(T t);

    public abstract <T> T deserialize (ByteBuffer bytes );

    public abstract <T> T deserialize(ByteBuffer bytes, ClassLoader loader);

    public abstract SerializationStream serializeStream(OutputStream s );

    public abstract DeserializationStream deserializeStream(InputStream s) ;

    public abstract DeserializationStream deserializeStream(InputStream s,ClassLoader classLoader) ;
}
abstract class SerializationStream {
    /** The most general-purpose method to write an object. */
    public abstract <T> SerializationStream writeObject(T t);
    /** Writes the object representing the key of a key-value pair. */
    public <T> SerializationStream  writeKey( T key) {
        return writeObject(key);
    }
    /** Writes the object representing the value of a key-value pair. */
    public <T> SerializationStream writeValue(T value){
        return writeObject(value);
    }
    public abstract void  flush();
    public abstract void  close();

    public <T> SerializationStream writeAll(Iterator<T> iter) {
        while (iter.hasNext()) {
            writeObject(iter.next());
        }
        return this;
    }
}
abstract class DeserializationStream {
    /**
     * The most general-purpose method to read an object.
     */
    public abstract <T> T readObject();

    /**
     * Reads the object representing the key of a key-value pair.
     */
    public <T> T readKey() {
        return readObject();
    }

    /**
     * Reads the object representing the value of a key-value pair.
     */
    public <T> T readValue() {
        return readObject();
    }

    public abstract void close();

    /**
     * Read the elements of this stream through an iterator. This can only be called once, as
     * reading each element will consume data from the input source.
     */
    public  <T> Iterator<T> asIterator(){
        return new NextIterator<T>() {
            @Override
            public T getNext() {
                try{
                    return readObject();
                }catch(Exception eof){
                    if(eof instanceof  EOFException){
                        finished = true ;
                    }
                    return null;
                }
            }

            @Override
            protected void close() {
                DeserializationStream.this.close();
            }
        };
    }

}